src/Psalm/Internal/Analyzer/ProjectAnalyzer.php

```diff
@@ -308,7 +308,11 @@ public function __construct(
         $file_extensions = $this->config->getFileExtensions();
 
         foreach ($this->config->getProjectDirectories() as $dir_name) {
-            $file_paths = $this->file_provider->getFilesInDir($dir_name, $file_extensions);
+            $file_paths = $this->file_provider->getFilesInDir(
+                $dir_name,
+                $file_extensions,
+                [$this->config, 'isInProjectDirs']
+            );
 
             foreach ($file_paths as $file_path) {
                 if ($this->config->isInProjectDirs($file_path)) {
```

Comment:
With the filter pushed down to the iterator, do we still need to check whether the returned file paths are in project / extra dirs?

Reply:
Currently - yes (the iterator returns `true` for all leaves), but the filter can be made more universal: #7201

Reply:
All _first-level_ leaves, that is. But yeah, got you.